### PR TITLE
[protobuf] further refactor of pb definitions into pkg/proto

### PR DESCRIPTION
**Pull request by truthbk**
### What does this PR do?

This PR continues the consolidation of protobuf definitions into the `pkg/proto` package. We were missing a few elements from the trace team, and some other miscellaneous changes (including some small tooling fixes). 

### Motivation

We want to make sure working with `pkg/proto` is consistent across the repository, and this PR hopes to cement the current protobuf workflow by continuing to move relevant bits and pieces into said `pkg/proto` sub-package. There are no functional aspirations in this PR beyond the developer experience and should add or remove no features. 

### Additional Notes

It's important to be thorough in the review of the protobuf changes given that we have moved quite a few protobuf messages around and that this PR bit-rotted for a while about 6 weeks before being re-worked on. Compilation success is a good indicator that things are fine, but I would really appreciate all code-owners signing off on this one.

The sooner we can get this to land, the better as these efforts can enter code-conflict situations rapidly. 

### Possible Drawbacks / Trade-offs

No trade-offs are expected.

### Describe how to test/QA your changes

There is no easy way to QA a relatively large refactor, but we should focus on evaluating the correctness and validity of all protobuf-based communications, be it grpc, or any serialized form of communication that may rely on this.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by brycekahle** [comment]
What the reason behind the `protodep` package, given the `agent-payload` repo supports go modules now?

**Comment by truthbk** [comment]
> What the reason behind the `protodep` package, given the `agent-payload` repo supports go modules now?

We don't just have `agent-payload` as a protobuf dependency, we have multiple other definitions we rely on for grpc, and others. Protodep is a nice tool to make sure those are tracked nicely. I'm not sure if I'm missing something from your question. Also, this PR does not introduce protodep, we have been using it for a while.

**Review by albertvaka — APPROVED** [review]

**Comment by brycekahle** [comment]
> Also, this PR does not introduce protodep, we have been using it for a while.

This PR does add the `agent-payload` to protodep. Since it wasn't necessary before, I'm curious what value it is providing?

**Comment by brycekahle** [comment]
> I'm not sure if I'm missing something from your question

I think the `dep` part of `protodep` was confusing me, especially since it uses the same TOML file format.

**Comment by djmitche** [comment]
I was hoping for some kind of README or other documentation to guide me as to what I'm looking at.  I'm also happy to just click "Approve" on the basis that @truthbk knows what he's doing :)

File: pkg/config/remote/client_test.go [hunk 1]
```diff
@@ -17,8 +17,8 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/config"
 	rdata "github.com/DataDog/datadog-agent/pkg/config/remote/data"
 	"github.com/DataDog/datadog-agent/pkg/config/remote/meta"
-	"github.com/DataDog/datadog-agent/pkg/proto/pbgo"
-	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	pbgocore "github.com/DataDog/datadog-agent/pkg/proto/pbgo/core"
```
Comment (vickenty) [inline pkg/config/remote/client_test.go:20]: Do you think it would make sense to prefix leaf package names with `pb` or `pbgo`, so that they don't need to be renamed every time we import them?

File: pkg/config/remote/client_test.go [hunk 2]
```diff
@@ -108,8 +108,8 @@ func TestClientValidResponse(t *testing.T) {
 
 	targetsKey := generateKey()
 	embeddedRoot := generateRoot(generateKey(), 1, targetsKey)
-	apmConfig := pb.APMSampling{
-		TargetTPS: []pb.TargetTPS{{Service: "service1", Env: "env1", Value: 4}},
+	apmConfig := pbgotrace.APMSampling{
+		TargetTps: []pbgotrace.TargetTPS{{Service: "service1", Env: "env1", Value: 4}},
```
Comment (vickenty) [inline pkg/config/remote/client_test.go:112]: Is there a reason to lowercase `TPS` in the field name here? IIRC go codestyle mandates uppercase for acronyms.

File: pkg/proto/datadog/trace/remote_rates.proto [hunk 3]
```diff
@@ -0,0 +1,25 @@
+syntax = "proto3";
```
Comment (vickenty) [inline pkg/proto/datadog/trace/remote_rates.proto:1]: This shows up as a new file, while others were moved. Is this git just failing to detect a rename?

File: pkg/proto/pbgo/core/remoteconfig_gen.go [hunk 4]
```diff
@@ -0,0 +1,5720 @@
+package core
+
+// Code generated by github.com/tinylib/msgp DO NOT EDIT.
```
Comment (vickenty) [inline pkg/proto/pbgo/core/remoteconfig_gen.go:3]: Why do we have messagepack impl in a protobuf directory?

File: pkg/trace/api/otlp.go [hunk 5]
```diff
@@ -20,13 +20,13 @@ import (
 	"sync"
 	"time"
 
+	pb "github.com/DataDog/datadog-agent/pkg/proto/pbgo/trace"
+	"github.com/DataDog/datadog-agent/pkg/proto/pbgo/trace/otlp"
```
Comment (vickenty) [inline pkg/trace/api/otlp.go:24]: It probably should be alised to `otlppb` or `pbgotlp` to avoid confusion with non-protobuf code. The former would let the patch be smaller too.

File: tasks/go.py [hunk 6]
```diff
@@ -403,13 +403,46 @@ def generate_licenses(ctx, filename='LICENSE-3rdparty.csv', verbose=False):
...
+    protodep_root = os.path.join(proto_root, "protodep")
 
     print(f"nuking old definitions at: {proto_root}")
-    file_list = glob.glob(os.path.join(proto_root, "pbgo", "*.go"))
+    file_list = glob.glob(os.path.join(proto_root, "pbgo", "*.pb.go"))
+    for file_path in file_list:
+        try:
+            os.remove(file_path)
+        except OSError:
+            print("Error while deleting file : ", file_path)
```
Comment (vickenty) [inline tasks/go.py]: Punctuation usually goes without space before, and the space after is added automatically by Python.

```suggestion
            print("Error while deleting file:", file_path)
```

File: pkg/trace/sampler/sampler.go [hunk 7]
```diff
@@ -53,6 +53,10 @@ const (
 // SamplingPriority is the type encoding a priority sampling decision.
 type SamplingPriority int8
 
+// SamplingMechanism this type is an alias for pkg/proto/datadog/trace/remote_rates.proto
```
Comment (vickenty) [inline pkg/trace/sampler/sampler.go]:
```suggestion
// SamplingMechanism is an alias for pkg/proto/datadog/trace/remote_rates.proto
```

This also removes some useful-looking comments from the original definition, we should probably keep them.

File: pkg/trace/sampler/sampler.go [hunk 8]
```diff
@@ -53,6 +53,10 @@ const (
...
+// TargetTPS.mechanism
+type SamplingMechanism uint32
```
Comment (vickenty) [inline pkg/trace/sampler/sampler.go]: Would it make sense to keep this definition together with TargetTPS as it used to be originally?

**Review by clamoriniere — APPROVED** [review]
/lgtm for files owned by `team/container-integrations

**Review by brycekahle — APPROVED** [review]
👍  for files owned by `agent-network`

**Review by robertjli — APPROVED** [review]
👍 for processes

**Comment by truthbk** [comment]
Serverless failures in github actions unrelated to this PR.

**Review by knusbaum — COMMENTED** [review]
Looks good from code review.
I'll work on some performance benchmarks before approving.

**Review by knusbaum — APPROVED** [review]
Looks good.

**Review by hmahmood — APPROVED** [review]

**Review by paulcacheux — APPROVED** [review]

**Review by ameske — APPROVED** [review]

File: pkg/otlp/internal/serializerexporter/consumer_test.go [hunk 9]
```diff
@@ -103,17 +105,14 @@ func TestConsumeAPMStats(t *testing.T) {
...
+	one := &pb.ClientStatsPayload{}
+	two := &pb.ClientStatsPayload{}
+	err := msgp.Decode(sc.apmstats[0], one)
 	require.NoError(t, err)
-	err = msgp.Decode(sc.apmstats[1], &two)
+	err = msgp.Decode(sc.apmstats[1], two)
 	require.NoError(t, err)
-	// We add back the tags to the originally added statsPayloads because that's
-	// what we need to compare against: ConsumeAPMStats adds the extraTags
-	statsPayloads[0].Tags = append(statsPayloads[0].Tags, sc.extraTags...)
```
Comment (songy23) [inline pkg/otlp/internal/serializerexporter/consumer_test.go:113]: Why is this line removed?

File: pkg/otlp/internal/serializerexporter/exporter.go [hunk 10]
```diff
@@ -176,8 +176,8 @@ func (e *exporter) ConsumeMetrics(ctx context.Context, ld pmetric.Metrics) error
 	}
 
 	consumer.addTelemetryMetric(e.hostname)
-	if rmt.HasRuntimeMetrics {
-		consumer.addRuntimeTelemetryMetric(e.hostname, rmt.LanguageTags)
+	if len(rmt.Languages) != 0 {
```
Comment (songy23) [inline pkg/otlp/internal/serializerexporter/exporter.go:179]: You don't need this if condition, `addRuntimeTelemetryMetric` won't do anything if `rmt.Languages` is empty.

**Review by songy23 — APPROVED** [review]
LGTM for opentelemetry

File: pkg/trace/writer/stats.go [hunk 11]
```diff
@@ -193,17 +196,24 @@ func (w *StatsWriter) buildPayloads(sp pb.StatsPayload, maxEntriesPerPayload int
...
 		nbBuckets = 0
+		current = &pb.StatsPayload{
+			AgentHostname:  sp.AgentHostname,
+			AgentEnv:       sp.AgentEnv,
+			AgentVersion:   sp.AgentVersion,
+			ClientComputed: sp.ClientComputed,
+		}
 	}
 	for _, p := range split {
+		fmt.Printf("Payload has %v entries, already storing %v entries\n", p.nbEntries, nbEntries)
```
Comment (hush-hush) [inline pkg/trace/writer/stats.go:209]: Should we keep this ?

File: pkg/trace/writer/stats.go [hunk 12]
```diff
@@ -193,17 +196,24 @@ func (w *StatsWriter) buildPayloads(sp pb.StatsPayload, maxEntriesPerPayload int
...
 		if nbEntries+p.nbEntries > maxEntriesPerPayload {
+			fmt.Printf("NEW PAYLOAD REQUIRED\n")
```
Comment (hush-hush) [inline pkg/trace/writer/stats.go:211]: Or this ?

File: pkg/trace/stats/client_stats_aggregator_test.go [hunk 13]
```diff
@@ -244,11 +256,30 @@ func TestFuzzCountFields(t *testing.T) {
...
+			actual = append(actual, s)
+		}
+		for _, s := range aggCounts.Stats[0].Stats[0].Stats {
+			if s == nil {
+				continue
+			}
+			expected = append(expected, s)
+		}
+		fmt.Printf("length actual: %v\n", len(actual))
+		fmt.Printf("length expected: %v\n", len(expected))
```
Comment (hush-hush) [inline pkg/trace/stats/client_stats_aggregator_test.go:280]: To be removed, no ?

File: pkg/proto/pbgo/trace/decoder_bytes_test.go [hunk 14]
```diff
@@ -86,5 +89,6 @@ func TestDecodeInvalidUTF8Bytes(t *testing.T) {
 	if _, err = got.UnmarshalMsg(bts); err != nil {
 		t.Fatal(err)
 	}
+	fmt.Printf("accept: %v\ngot: %v\n", accept, got)
```
Comment (hush-hush) [inline pkg/proto/pbgo/trace/decoder_bytes_test.go:92]: Debug log ?

File: pkg/proto/pbgo/trace/decoder_bytes_test.go [hunk 15]
```diff
@@ -61,19 +63,20 @@ func TestDecodeBytes(t *testing.T) {
 	if _, err = got.UnmarshalMsg(bts); err != nil {
 		t.Fatal(err)
 	}
+	fmt.Printf("want: %v\ngot: %v\n", want, got)
```
Comment (hush-hush) [inline pkg/proto/pbgo/trace/decoder_bytes_test.go:66]: Debug log ?

**Review by hush-hush — APPROVED** [review]

**Comment by truthbk** [comment]
After 9 approvals, and at the rate this is bit-rotting, I am going to merge with privileges. Failures on tests at the time of this merge can also be seen on `main`.
